File: crossecore-csharp/Java/Java/Java_/MethodDeclarationBase.cs
/* CrossEcore is a cross-platform modeling framework that generates C#, TypeScript, 
 * JavaScript, Swift code from Ecore models with embedded OCL (http://www.crossecore.org/).
 * The original Eclipse Modeling Framework is available at https://www.eclipse.org/modeling/emf/.
 * 
 * contributor: Simon Schwichtenberg
 */
 
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using oclstdlib;
using Ecore;
namespace Java_{
	public class MethodDeclarationBase 
	:AbstractMethodDeclarationImpl, MethodDeclaration
	{
		private int _extraArrayDimensions = 0;
		public virtual int extraArrayDimensions
		{
		get { 
			return _extraArrayDimensions;
		}
		set { _extraArrayDimensions = value; }
		}
		private TypeAccess _returnType;
		public virtual TypeAccess returnType
		{
			get {
			
				return _returnType;
			}
			set {
				if (value != _returnType) {
					NotificationChain msgs = null;
					if (_returnType != null){
						msgs = ((InternalEObject)_returnType).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.METHODDECLARATION_RETURNTYPE, null, msgs);
					}
					if (value != null){
						msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.METHODDECLARATION_RETURNTYPE, null, msgs);
					}
					msgs = basicSetReturnType(value, msgs);
					if (msgs != null) {
						msgs.dispatch();
					}
				}
				else if (eNotificationRequired()){
					eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.METHODDECLARATION_RETURNTYPE , value, value));
				}
				}
		}
		private MethodDeclaration _redefinedMethodDeclaration;
		public virtual MethodDeclaration redefinedMethodDeclaration
		{
			get {
			
				return _redefinedMethodDeclaration;
			}
			set {
				if (value != _redefinedMethodDeclaration) {
					NotificationChain msgs = null;
					if (_redefinedMethodDeclaration != null){
						msgs = ((InternalEObject)_redefinedMethodDeclaration).eInverseRemove(this, Java_PackageImpl.METHODDECLARATION_REDEFINITIONS, typeof(MethodDeclaration), msgs);
					}
					if (value != null){
						msgs = ((InternalEObject)value).eInverseAdd(this, Java_PackageImpl.METHODDECLARATION_REDEFINITIONS, typeof(MethodDeclaration), msgs);
					}
					msgs = basicSetRedefinedMethodDeclaration(value, msgs);
					if (msgs != null) {
						msgs.dispatch();
					}
				}
				else if (eNotificationRequired()){
					eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION , value, value));
				}
				}
		}
		private Set<MethodDeclaration> _redefinitions;
		
		public virtual Set<MethodDeclaration> redefinitions
		{
			get {
				if(_redefinitions==null){
					_redefinitions = new Set<MethodDeclaration>(this, Java_PackageImpl.METHODDECLARATION_REDEFINITIONS, Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION);
				}
				return _redefinitions;
			}
		
		}
		
		protected override EClass eStaticClass() {
			return Java_PackageImpl.Literals.METHODDECLARATION;
		}
		
		public override NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.METHODDECLARATION_REDEFINITIONS:
					return redefinitions.basicAdd((MethodDeclaration)otherEnd, msgs);
				case Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION:
					if (_redefinedMethodDeclaration != null){
						msgs = ((InternalEObject)_redefinedMethodDeclaration).eInverseRemove(this, Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION, typeof(MethodDeclaration), msgs);
					}
					return basicSetRedefinedMethodDeclaration((MethodDeclaration)otherEnd, msgs);
			}
			return base.eInverseAdd(otherEnd, featureID, msgs);
		}
		
		public override NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.METHODDECLARATION_REDEFINITIONS:
					return redefinitions.basicRemove((MethodDeclaration)otherEnd, msgs);
				case Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION:
					return basicSetRedefinedMethodDeclaration(null, msgs);
			}
			return base.eInverseRemove(otherEnd, featureID, msgs);
		}
		
		public NotificationChain basicSetRedefinedMethodDeclaration(MethodDeclaration newobj, NotificationChain msgs) {
			var oldobj = _redefinedMethodDeclaration;
			_redefinedMethodDeclaration = newobj;
			if (eNotificationRequired()) {
				var notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION, oldobj, newobj);
				if (msgs == null){
					msgs = notification;
				}
				else{
					msgs.add(notification);
				}
			}
			return msgs;
		}
		public NotificationChain basicSetReturnType(TypeAccess newobj, NotificationChain msgs) {
			var oldobj = _returnType;
			_returnType = newobj;
			if (eNotificationRequired()) {
				var notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.METHODDECLARATION_RETURNTYPE, oldobj, newobj);
				if (msgs == null){
					msgs = notification;
				}
				else{
					msgs.add(notification);
				}
			}
			return msgs;
		}
		
		public override object eGet(int featureID, bool resolve, bool coreType) {
			switch (featureID) {
				case Java_PackageImpl.METHODDECLARATION_EXTRAARRAYDIMENSIONS:
					return extraArrayDimensions;
				case Java_PackageImpl.METHODDECLARATION_RETURNTYPE:
					return returnType;
				case Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION:
					return redefinedMethodDeclaration;
				case Java_PackageImpl.METHODDECLARATION_REDEFINITIONS:
					return redefinitions;
			}
			return base.eGet(featureID, resolve, coreType);
		}
		
		
		public override void eSet(int featureID, object newValue) {
			switch (featureID) {
				case Java_PackageImpl.METHODDECLARATION_EXTRAARRAYDIMENSIONS:
					extraArrayDimensions = (int) newValue;
					return;
				case Java_PackageImpl.METHODDECLARATION_RETURNTYPE:
					returnType = (TypeAccess) newValue;
					return;
				case Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION:
					redefinedMethodDeclaration = (MethodDeclaration) newValue;
					return;
				case Java_PackageImpl.METHODDECLARATION_REDEFINITIONS:
					redefinitions.Clear();
					redefinitions.AddRange(((List<EObject>)newValue)?.Cast<MethodDeclaration>());
					return;
			}
			base.eSet(featureID, newValue);
		}
		
		/*
		public override bool eIsSet(int featureID) {
			switch (featureID) {
				case Java_PackageImpl.METHODDECLARATION_EXTRAARRAYDIMENSIONS:
					return _extraArrayDimensions != EXTRAARRAYDIMENSIONS_EDEFAULT;
				case Java_PackageImpl.METHODDECLARATION_RETURNTYPE:
					return _returnType != null; //single != null;
				case Java_PackageImpl.METHODDECLARATION_REDEFINEDMETHODDECLARATION:
					return getRedefinedMethodDeclaration() != null; //single, volatile
				case Java_PackageImpl.METHODDECLARATION_REDEFINITIONS:
					return getRedefinitions().isEmpty(); //many, volatile
			}
			return base.eIsSet(featureID);
		}
		*/
		
		
		
	}
}

File: crossecore-java/src/java_/MethodDeclarationBase.java
package java_;
import org.eclipse.emf.common.notify.Notification;
import org.eclipse.emf.common.notify.NotificationChain;
import org.eclipse.emf.common.notify.impl.NotificationImpl;
import org.eclipse.emf.ecore.InternalEObject;
import org.eclipse.emf.ecore.*;
import org.eclipse.emf.ecore.impl.ENotificationImpl;
public class MethodDeclarationBase 
extends AbstractMethodDeclarationImpl implements MethodDeclaration
{
	protected static final int EXTRAARRAYDIMENSIONS_EDEFAULT = 0;
	private int _extraArrayDimensions = EXTRAARRAYDIMENSIONS_EDEFAULT;
	public int getExtraArrayDimensions()
	{
		return _extraArrayDimensions;
	}
	public void setExtraArrayDimensions(int value){
		
		int oldValue = _extraArrayDimensions;
		_extraArrayDimensions = value;
		if (eNotificationRequired())
			eNotify(new ENotificationImpl(this, Notification.SET, Java_PackageImpl.METHOD_DECLARATION__EXTRA_ARRAY_DIMENSIONS, oldValue, value));
		
	}
	
	
	
	private TypeAccess _returnType;
	public TypeAccess getReturnType()
	{
		
			return _returnType;
		
		
	}
	public void setReturnType(TypeAccess value){
		if (value != _returnType) {
			NotificationChain msgs = null;
			if (_returnType != null){
				msgs = ((InternalEObject)_returnType).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.METHOD_DECLARATION__RETURN_TYPE, null, msgs);
			}
			if (value != null){
				msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.METHOD_DECLARATION__RETURN_TYPE, null, msgs);
			}
			msgs = basicSetReturnType(value, msgs);
			if (msgs != null) {
				msgs.dispatch();
			}
		}
		else if (eNotificationRequired()){
			eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.METHOD_DECLARATION__RETURN_TYPE , value, value));
		}
	}
	private MethodDeclaration _redefinedMethodDeclaration;
	public MethodDeclaration getRedefinedMethodDeclaration()
	{
		
			return _redefinedMethodDeclaration;
		
		
	}
	public void setRedefinedMethodDeclaration(MethodDeclaration value){
		if (value != _redefinedMethodDeclaration) {
			NotificationChain msgs = null;
			if (_redefinedMethodDeclaration != null){
				msgs = ((InternalEObject)_redefinedMethodDeclaration).eInverseRemove(this, Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS, MethodDeclaration.class, msgs);
			}
			if (value != null){
				msgs = ((InternalEObject)value).eInverseAdd(this, Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS, MethodDeclaration.class, msgs);
			}
			msgs = basicSetRedefinedMethodDeclaration(value, msgs);
			if (msgs != null) {
				msgs.dispatch();
			}
		}
		else if (eNotificationRequired()){
			eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION , value, value));
		}
	}
	private com.crossecore.ocl.Set<MethodDeclaration> _redefinitions;
	
	public com.crossecore.ocl.Set<MethodDeclaration> getRedefinitions()
	{
		if(_redefinitions==null){
			_redefinitions = new com.crossecore.ocl.Set<MethodDeclaration>(MethodDeclaration.class, this, Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS, Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION);
		}
		return _redefinitions;
	
	}

	
	@Override
	protected EClass eStaticClass() {
		return Java_PackageImpl.Literals.METHODDECLARATION;
	}
	
	@Override
	public NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION:
				if (_redefinedMethodDeclaration != null){
					msgs = ((InternalEObject)_redefinedMethodDeclaration).eInverseRemove(this, Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION, MethodDeclaration.class, msgs);
				}
				return basicSetRedefinedMethodDeclaration((MethodDeclaration)otherEnd, msgs);
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS:
				return getRedefinitions().basicAdd((MethodDeclaration)otherEnd, msgs);
		}
		return super.eInverseAdd(otherEnd, featureID, msgs);
	}
	
	@Override
	public NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION:
				return basicSetRedefinedMethodDeclaration(null, msgs);
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS:
				return getRedefinitions().basicRemove((MethodDeclaration)otherEnd, msgs);
		}
		return super.eInverseRemove(otherEnd, featureID, msgs);
	}
	
	public NotificationChain basicSetRedefinedMethodDeclaration(MethodDeclaration newobj, NotificationChain msgs) {
		MethodDeclaration oldobj = _redefinedMethodDeclaration;
		_redefinedMethodDeclaration = newobj;
		if (eNotificationRequired()) {
			ENotificationImpl notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION, oldobj, newobj);
			if (msgs == null){
				msgs = notification;
			}
			else{
				msgs.add(notification);
			}
		}
		return msgs;
	}
	public NotificationChain basicSetReturnType(TypeAccess newobj, NotificationChain msgs) {
		TypeAccess oldobj = _returnType;
		_returnType = newobj;
		if (eNotificationRequired()) {
			ENotificationImpl notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.METHOD_DECLARATION__RETURN_TYPE, oldobj, newobj);
			if (msgs == null){
				msgs = notification;
			}
			else{
				msgs.add(notification);
			}
		}
		return msgs;
	}
	
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
			case Java_PackageImpl.METHOD_DECLARATION__EXTRA_ARRAY_DIMENSIONS:
				return getExtraArrayDimensions();
			case Java_PackageImpl.METHOD_DECLARATION__RETURN_TYPE:
				return getReturnType();
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION:
				return getRedefinedMethodDeclaration();
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS:
				return getRedefinitions();
		}
		return super.eGet(featureID, resolve, coreType);
	}
	
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
			case Java_PackageImpl.METHOD_DECLARATION__EXTRA_ARRAY_DIMENSIONS:
				setExtraArrayDimensions((int) newValue);
				return;
			case Java_PackageImpl.METHOD_DECLARATION__RETURN_TYPE:
				setReturnType((TypeAccess) newValue);
				return;
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION:
				setRedefinedMethodDeclaration((MethodDeclaration) newValue);
				return;
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS:
				getRedefinitions().clear();
				getRedefinitions().addAll((java.util.Collection<? extends MethodDeclaration>) newValue);
				return;
		}
		super.eSet(featureID, newValue);
	}
	
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
			case Java_PackageImpl.METHOD_DECLARATION__EXTRA_ARRAY_DIMENSIONS:
				return _extraArrayDimensions != EXTRAARRAYDIMENSIONS_EDEFAULT;
			case Java_PackageImpl.METHOD_DECLARATION__RETURN_TYPE:
				return _returnType != null; //single != null;
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINED_METHOD_DECLARATION:
				return getRedefinedMethodDeclaration() != null; //single, volatile
			case Java_PackageImpl.METHOD_DECLARATION__REDEFINITIONS:
				return getRedefinitions().isEmpty(); //many, volatile
		}
		return super.eIsSet(featureID);
	}
	
	
	
}
